### ISSUE TITLE
LeakyRelu support in contrib.quantize

### ISSUE THREAD

**Issue by myworldraj** [tensorflow/tensorflow#27996]
**System information**
- TensorFlow version (you are using): 1.13.1
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**
I am trying to create an eval graph with fake quantization nodes, from a checkpoint file, of a tensorflow implementation of YoloV3. My goal is to eventually convert this to a UINT8 (fully quantized) .tflite inference model using TOCO.

My code snippet is like the following -

        with tf.variable_scope('yolov3'):
            boxes, confs, probs = model.forward(inputs, is_training=False)

        scores = confs * probs
        print("=>", boxes.name[:-2], scores.name[:-2])
        cpu_out_node_names = [boxes.name[:-2], scores.name[:-2]]
        saver = tf.train.Saver(var_list=tf.global_variables(scope='yolov3'))

        saver.restore(sess, flags.ckpt_file)
        tf.contrib.quantize.create_eval_graph(input_graph=graph)
        sess.run(tf.global_variables_initializer())
        print('=> checkpoint file restored from ', flags.ckpt_file)
        utils.freeze_graph(sess, './experiment/yolov3_FQ.pb', cpu_out_node_names)

The tf.contrib.quantize.create_eval_graph(...) call itself finishes with no error, but when I run TOCO to convert the graph to UINT8 tflite, it complains that LeakyRelu doesn't have min/max parameters.

Following is my TOCO command -

bazel --output_user_root=/local/mnt/workspace/rganguly/deep_learning/bazel_new/build run --config=opt tensorflow/lite/toco:toco -- \
--output_file=/local/mnt/workspace/rganguly/deep_learning/live/yolov3/yolov3_.tflite \
--input_file=/local/mnt/workspace/rganguly/deep_learning/live/yolov3/tensorflow-yolov3/experiment/yolov3_fake_quant.pb \
--input_shapes=1,416,416,3 \
--input_arrays=Placeholder \
--output_arrays='yolov3/concat_3','mul' \
--inference_type=QUANTIZED_UINT8 \
--mean_values=128 \
--std_dev_values=127 \

And following is the error I get -

2019-04-19 16:56:59.708619: I tensorflow/lite/toco/graph_transformations/graph_transformations.cc:39] Before Removing unused ops: 1613 operators, 2394 arrays (0 quantized)
2019-04-19 16:56:59.747329: I tensorflow/lite/toco/graph_transformations/graph_transformations.cc:39] Before general graph transformations: 1613 operators, 2394 arrays (0 quantized)
2019-04-19 16:57:06.761701: I tensorflow/lite/toco/graph_transformations/graph_transformations.cc:39] After general graph transformations pass 1: 286 operators, 471 arrays (1 quantized)
2019-04-19 16:57:06.767392: I tensorflow/lite/toco/graph_transformations/graph_transformations.cc:39] Before pre-quantization graph transformations: 286 operators, 471 arrays (1 quantized)
2019-04-19 16:57:06.770164: I tensorflow/lite/toco/graph_transformations/graph_transformations.cc:39] After pre-quantization graph transformations pass 1: 211 operators, 396 arrays (1 quantized)
2019-04-19 16:57:06.772654: I tensorflow/lite/toco/graph_transformations/graph_transformations.cc:39] Before Group bidirectional sequence lstm/rnn: 211 operators, 396 arrays (1 quantized)
2019-04-19 16:57:06.775354: F tensorflow/lite/toco/tooling_util.cc:1708] Array yolov3/darknet-53/Conv/LeakyRelu, which is an input to the Pad operator producing the output array yolov3/darknet-53/Pad, is lacking min/max data, which is necessary for quantization. If accuracy matters, either target a non-quantized output format, or run quantized training with your model from a floating point checkpoint to change the input graph to contain min/max information. If you don't care about accuracy, you can pass --default_ranges_min= and --default_ranges_max= for easy experimentation.

**Will this change the current api? How?**
- NO

**Who will benefit with this feature?**
- Anyone who has a LeakyRelu in the tensorflow graph and wants to transform it to a fully quantized/UINT8 .tflite model. In other words, YoloV3 cannot be converted to a UINT8 .tflite model without this support.

**Any Other info.**
There are multiple other issues I faced in TOCO while converting this model, but I will summarize them in another ticket.

**Comment by amitsrivastava78** [comment]
@myworldraj , i think the problem is that quantization support for LeakyRelu is not supported in tflite, i have raised on PR for the same #27061, kindly test with this PR and let me know if you were able to succeed.

Regards
Amit

**Comment by achandraa** [comment]
@myworldraj : Please have a look and let us know if that helps.Thanks!

**Comment by myworldraj** [comment]
@amitsrivastava78, there are two issues here -

1. Lack of LeakyRelu support in tf.contrib.quantize.create_training_graph(...) and tf.contrib.quantize.create_eval_graph(...), which in turn calls quantize.Quantize(), for which **this ticket has been raised**

2. Lack of quantized LeakyRelu support in the TFLite inference path, for which I had already found your commit, and have integrated it in my workspace.

To help you understand in easier terms - we can **only** use your commit, once we first create a **quantization friendly** graph in tensorflow. That can only happen, when we add support of LeakyRelu in Quantize(). Then TOCO or tflite_convert can correctly convert it to a fully quantized (i.e. UINT8) .tflite model, That is what I am looking for as a resolution to this ticket.

I hope this clarifies.

**Comment by amitsrivastava78** [comment]
@myworldraj , thanks for the clarification, I would love to help you out can you please share the code and other details, so i can simulate it on my workstation and give the solution to you ASAP.

Regards
Amit

**Comment by myworldraj** [comment]
@amitsrivastava78,

I am using tensorflow 1.13.1, built from source.

Block A steps
=================================
- Download the source code from https://github.com/YunYang1994/tensorflow-yolov3
- Use the modified file: convert_weight_raj.txt (rename to .py)
[convert_weight_raj.txt](https://github.com/tensorflow/tensorflow/files/3113557/convert_weight_raj.txt)
- Use the modified file: yolov3.txt (rename it to .py)
[yolov3.txt](https://github.com/tensorflow/tensorflow/files/3113564/yolov3.txt)
- run python3 convert_weight_raj.py --freeze, which will save the .pb with fake quant nodes.
=================================

Block B steps
=================================
- run TOCO, like 
bazel  run --config=opt tensorflow/lite/toco:toco -- \
--output_file=yolov3.tflite \
--input_file=yolov3_FQ.pb \
--input_shapes=1,416,416,3 \
--input_arrays=Placeholder \
--output_arrays='yolov3/concat_3','mul' \
--inference_type=QUANTIZED_UINT8 \
--mean_values=128 \
--std_dev_values=127 \

- When you run TOCO first time, you will see multiple errors reported by TOCO, because there are many issues with TOCO. I did the following steps to work around with most issues -

  1. Integrated your commit
  2. Made changes to ignore lack of support for Cast operator (basically made it pass-through)
  3. Added SplitV and LeakyRelu to the list of supported quantizable ops

- I am attaching a list of files I changed, in addition to integrating your commit, to make TOCO work -

[TOCO_changes.zip](https://github.com/tensorflow/tensorflow/files/3113635/TOCO_changes.zip)

- After you rebuild TOCO and run it again as above, you will come to the **real issue** of this ticket, which is - TOCO will complain that LeakyRelu doesn't have a min/max. 

- So we will need to go back to the drawing board (to tensorflow/contrib/quantize/python/), and figure out how to add support for LeakyRelu activation, such that it satisfies TOCO

**Comment by amitsrivastava78** [comment]
@myworldraj thanks for sharing the details, was able to reproduce the error, but since i will be on travel till next Wednesday, it might take some time for me to comeback with solution for you.

Regards
Amit

**Comment by myworldraj** [comment]
@amitsrivastava78,

Wanted to give you a few more inputs - I did a quick experimentation and added LeakyRelu to the list of supported activation-s in the following places in tensorflow/contrib/quantize/python/quantize.py -
- _ACTIVATION_TYPES = {'Relu', 'Relu6', 'LeakyRelu', 'Identity'}
- _RELU_TYPES = {'Relu', 'Relu6', 'LeakyRelu'}
- _VALID_ACTIVATION_OP = {'Relu', 'Relu6', 'LeakyRelu'}

Now, TOCO is happy (at least didn't complain) and I get proper FakeQuant... nodes after LeakyRelu. But now there is another problem - since TOCO cannot possibly fuse the LeakyReuly activation with the preceding op, the Add (which is a preceding op to LeakyRelu in this graph) is dangling without a min/max and TOCO is erroring out.

Looks like we need to take a comprehensive look on how to insert FakeQuant nodes to infusable activations, and make wholesome changes in both the Quantize pass and the TOCO pass.

**Comment by myworldraj** [comment]
Is there any update on this issue?

**Comment by liyunlu0618** [comment]
We don't plan to add more new features to the quantization rewriter. Please stay tuned for our new keras based quantization aware training API.

If you want to get a fully quantized model, you can take a look at our post-training integer quantization for now:
https://medium.com/tensorflow/tensorflow-model-optimization-toolkit-post-training-integer-quantization-b4964a1ea9ba

Also, there has been a lot of asks for LeakyRelu quantization support. We'll work on that shortly. Thanks!

**Comment by willbattel** [comment]
@liyunlu0618 do you have any links/info you can provide regarding that new Keras API, or where we can follow its development if not yet available? We're trying to do quantization-aware training in 2.0, where the `tf.contrib.quantize` functions are no longer available.
